Distant sensor array

schunkes opened the pull request:
# Description

This PR will introduce a variant of the `DistantMeasure`, that lets users specify a list of directions, which will be treated individually, instead of defining a hemisphere or plane to record leaving radiance on.

There are open questions which must be addressed before this code can be merged:

- The `_postprocess_add_directions` method in `DistantArrayMeasure` adds a new coordinate to the dataset, which lets should let users select the results based on the direction they specified for a sensor. However xarray does not support the use of lists as coordinate entries, so for the moment I add the directions as strings, which is at least informative.
- Since I cannot use xarray's builting averaging methods here, I have to select the correct subset of results along the 'sensor_id' dimension and use `numpy.average`. But this leads to some issues with squeezed dimensions and I suspect it will be problematic if the `spectral_coordinate` dimension is not a scalar.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

leroyvn commented:
Hi @schunkes, a couple of general comments in addition to answers to your questions.

## Sensor direction coordinates

You can keep the sensor dimension linear, make it a simple integer, and add direction data as non-dimension coordinates, quite similar to what is done for `DistantMeasure`. We can then provide helpers to generate optimised direction sets and repack the data in order to make it sliceable (see *Side notes*).

## Dimension issues

I think the best is to allocate a dimension specifically dedicated to SPP split. You'll therefore have 2 dimensions varying sensor ID: a proper `sensor_id` and a `sensor_id_spp` dimension. Collapsing the second dimension should then be easy. In order to do so, you'll have to modify the dataset. I can see two possibilities:

- update the low-level dataset format (not great because it means that the most raw data we output is already processed);
- reindex the data prior to collapsing the SPP dimension (probably better, a little more 'arcane' but not too difficult to experiment thanks to the full access you have to the raw data).

The final assignment would still be something like:

```python
result_aggregated = result.assign(
    {"raw": result["raw"].weighted(weights).mean(dim="sensor_id_spp")},
)
```

## Side notes

(We'll take care of those after this prototype is ready.)

- I think we'll make this measure series the default distant measure and rename the hemispherical one `HDistant*Measure`.
- We'll have to provide a few helpers for users to easily configure directions and format results (*e.g.* plane, ring, hemisphere). How this will be implemented is to be discussed.

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+from ...units import unit_context_kernel as uck
+
+
+@measure_factory.register(type_id="distantarray")
```
Consistency: I'd rather have this called `distant_array` (better correspondence between camel and snake case).

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+    def _directions_validator(self, attribute, value):
+        if value.shape[1] != 3:
+            raise ValueError(
+                f"While initializing {attribute}: "
```
```suggestion
                f"While validating {attribute.name}: "
```
(validators can also be run upon setter call)

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+            )
+
+    def sensor_infos(self) -> List[SensorInfo]:
+        """
+        Return a tuple of sensor information data structures.
+
+        Returns → list[:class:`.SensorInfo`]:
+            List of sensor information data structures.
+        """
```
No need to repeat the docstring if it is inherited. However, it would be nice if the parent method  would document the rationale underlying sensor ID definition.

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+            {
+                "directions": (
+                    "sensor_id",
+                    [str(direction) for direction in self.directions],
```
IMO the best would be to store viewing angle values as non-dimension coordinates (see my general comment).

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+        return (1, 1)
+
+
+@measure_factory.register(type_id="distantradiancearray")
```
```suggestion
@measure_factory.register(type_id="distant_array_radiance")
```

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+@measure_factory.register(type_id="distantradiancearray")
+@parse_docs
+@attr.s
+class DistantRadianceArrayMeasure(DistantArrayMeasure):
```
```suggestion
class DistantArrayRadianceMeasure(DistantArrayMeasure):
```

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+@attr.s
+class DistantRadianceArrayMeasure(DistantArrayMeasure):
+    """
+    Distant radiance array measure scene element [``distantradiancearray``].
```
```suggestion
    Distant radiance array measure scene element [``distant_array_radiance``].
```

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+@measure_factory.register(type_id="distantreflectancearray")
+@parse_docs
+@attr.s
+class DistantReflectanceArrayMeasure(DistantArrayMeasure):
```
```suggestion
class DistantArrayReflectanceMeasure(DistantArrayMeasure):
```

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+        return ds
+
+
+@measure_factory.register(type_id="distantreflectancearray")
```
```suggestion
@measure_factory.register(type_id="distant_array_reflectance")
```

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+@attr.s
+class DistantReflectanceArrayMeasure(DistantArrayMeasure):
+    """
+    Distant reflectance array measure scene element [``distantreflectancearray``].
```
```suggestion
    Distant reflectance array measure scene element [``distant_array_reflectance``].
```

leroyvn commented on eradiate/scenes/measure/_distantsensorarray.py:
```diff
@@ -0,0 +1,218 @@
+@measure_factory.register(type_id="distantarray")
+@parse_docs
+@attr.s
+class DistantArrayMeasure(DistantMeasure):
```
Isn't this class supposed to be an interface definition only? It should not be registered to the factory and should inherit from `ABC` (just to mention explicitly that it is an abstract class).
```suggestion
class DistantArrayMeasure(DistantMeasure, ABC):
```

leroyvn reviewed (changes requested):
A few more focused comments. I guess the tests aren't aligned yet (you have a `measure_type` kwarg which I didn't see in your lib code).

leroyvn commented:
> 
> 
> I have an idea for the averaging, but it does sound a bit hacky: Instead of extracting the relevant results for one sensor, those which should be averaged for spp aggregation, we could create a padded `weights` array, which holds the corresponding spp value for all relevant sensors and 0 otherwise. In that case we could use the xarray mean method and would not have to bother with the other dimensions. 🤔
> 
> EDIT: Creating the padded `weights` is actually very easy with the `xr.where()` method. I am just not sure how to properly merge my results with the Dataset afterwards...

Just saw this: extending the weights array is maybe another solution, I hadn't though about it. I'll let you ponder options, we can discuss this later on ;)

schunkes commented:
I still haven't addressed your comments but I managed to get the raw data aggregation and assembly of the new Dataset to work :) Would you mind taking a look at it?

leroyvn commented on eradiate/scenes/measure/_core.py:
```diff
@@ -466,20 +465,55 @@ def to_dataset(self, aggregate_spps: bool = False) -> xr.Dataset:
+                .mean(dim="sensor_id")
+                .expand_dims({"sensor_id": [id_sensor]}, axis=1)
+            )
+            # spps themselves must not be scaled, but just masked to include
+            # the right values in the sum
+            weights_spp = xr.where(result.coords["sensor_id"].isin(relevant_ids), 1, 0)
+            spp_aggregated.append(
+                result.data_vars["spp"]
+                .weighted(weights_spp)
+                .sum(dim="sensor_id")
+                .expand_dims({"sensor_id": [id_sensor]}, axis=1)
+            )
```
```suggestion

            # Mask spp values so as to only include selected values
            # The sensor_id coordinate is redefined to strip the spp suffix
            weights_spp = xr.where(result.coords["sensor_id"].isin(sensor_ids), 1, 0)
            spp_aggregated.append(
                result.data_vars["spp"]
                .weighted(weights_spp)
                .sum(dim="sensor_id")
                .expand_dims({"sensor_id": [sensor_id_prefix]}, axis=1)
            )
```

leroyvn commented on eradiate/scenes/measure/_core.py:
```diff
@@ -466,20 +465,55 @@ def to_dataset(self, aggregate_spps: bool = False) -> xr.Dataset:
+                .expand_dims({"sensor_id": [id_sensor]}, axis=1)
+            )
+
+        raw_redone = xr.combine_by_coords(raw_aggregated)
+        spp_redone = xr.combine_by_coords(spp_aggregated)
```
```suggestion
        raw = xr.combine_by_coords(raw_aggregated)["raw"]
        spp = xr.combine_by_coords(spp_aggregated)["spp"]
```

leroyvn commented on eradiate/scenes/measure/_core.py:
```diff
@@ -466,20 +465,55 @@ def to_dataset(self, aggregate_spps: bool = False) -> xr.Dataset:
+        raw_redone = xr.combine_by_coords(raw_aggregated)
+        spp_redone = xr.combine_by_coords(spp_aggregated)
+
+        result_aggregated = xr.Dataset(
+            data_vars={
+                "raw": raw_redone["raw"],
+                "spp": spp_redone["spp"],
+                "sensor_id": list(measure_ids_sensor),
+            }
```
```suggestion
        result_aggregated = xr.Dataset(
            data_vars={
                "raw": raw,
                "spp": spp,
                "sensor_id": sensor_id_prefixes,
            }
```
(`sorted()` already returns a list)

leroyvn reviewed (commented):
Hi @schunkes, I made suggestions to clarify the logic of your aggregation code (I hope I understood correctly!). It's quite comprehensive IMO, so we can go with that.

schunkes commented:
Here we go, I included your feedback and I added a *from_angles* constructor.

leroyvn reviewed (changes requested):
Hi @schunkes, that's nice. I have a couple of comments. 

I also would like you to make some changes to the aggregation code (if not the ones I suggested, something in a similar spirit) because I'm quite sure that if I go through that code in 6 months, it will take me much more effort to remember the logics than if variables are named more consistently and comments a little more detailed.

leroyvn commented on eradiate/scenes/measure/_distant_array.py:
```diff
@@ -0,0 +1,221 @@
+from ...units import unit_context_kernel as uck
+
+
+@measure_factory.register(type_id="distant_array")
```
This is an ABC: it's not supposed to be registered to the factory.
```suggestion
```

leroyvn commented on eradiate/scenes/measure/_distant_array.py:
```diff
@@ -0,0 +1,221 @@
+    Distant sensor array measure scene element [``distantarray``].
+
+    This measure scene element is a thin wrapper around the ``DistantSensor``
+    measures. It records either the radiance or reflectance along the directions given by the
+    ``directions`` parameter.
```
This class technically doesn't: only its concrete child classes do.

leroyvn commented on eradiate/scenes/measure/_distant_array.py:
```diff
@@ -0,0 +1,221 @@
+        return (1, 1)
+
+
+@measure_factory.register(type_id="distant_array_radiance")
```
Let's add an alias:
```suggestion
@measure_factory.register(type_id="distant_array")
@measure_factory.register(type_id="distant_array_radiance")
```

leroyvn commented on eradiate/scenes/measure/_distant_array.py:
```diff
@@ -0,0 +1,221 @@
+@attr.s
+class DistantArrayRadianceMeasure(DistantArrayMeasure):
+    """
+    Distant array radiance measure scene element [``distant_array_radiance``].
```
```suggestion
    Distant array radiance measure scene element [``distant_array``, ``distant_array_radiance``].
```

leroyvn commented on eradiate/scenes/measure/_distant_array.py:
```diff
@@ -0,0 +1,221 @@
+    """
+    Distant array radiance measure scene element [``distant_array_radiance``].
+
+    This measure scene element is a thin wrapper around the ``DistantSensor``
+    measures. It records either the radiance or reflectance along the directions given by the
+    ``directions`` parameter.
```
```suggestion
    This measure scene element is a thin wrapper around the ``DistantSensor``
    measures. It records radiance along the directions given by the ``directions`` 
    parameter.
```

leroyvn commented on eradiate/scenes/measure/_distant_array.py:
```diff
@@ -0,0 +1,221 @@
+    """
+    Distant array reflectance measure scene element [``distant_array_reflectance``].
+
+    This measure scene element is a thin wrapper around the ``DistantSensor``
+    measures. It records either the radiance or reflectance along the directions given by the
+    ``directions`` parameter.
```
```suggestion
    This measure scene element is a thin wrapper around the ``DistantSensor``
    measures. It records the radiance and reflectance along the directions given by the
    ``directions`` parameter.
```

schunkes commented:
Alright, take another look. I think the dict interface is good, the keywords feel natural and fitting to me.

leroyvn commented on eradiate/scenes/measure/_core.py:
```diff
@@ -466,20 +465,58 @@ def to_dataset(self, aggregate_spps: bool = False) -> xr.Dataset:
-        result_aggregated = result_aggregated.drop_dims("sensor_id")
 
-        # Copy metadata (lost during aggregation)
         for var in list(result_aggregated.data_vars) + list(result_aggregated.coords):
```
```suggestion
        # Copy metadata
        for var in list(result_aggregated.data_vars) + list(result_aggregated.coords):
```
